pkg/pool-manager/virtualmachine_pool.go

```diff
@@ -445,51 +540,123 @@ func (p *PoolManager) vmWaitingCleanupLook() {
+		return nil, errors.New("failed to extract vm namespace and name")
+	}
+
+	requestUrl := fmt.Sprintf("apis/kubevirt.io/v1/namespaces/%s/virtualmachines/%s", vmNamespace, vmName)
+	log.V(1).Info("getvmInstance", "requestURI", requestUrl)
+
+	result := p.kubeClient.ExtensionsV1beta1().RESTClient().Get().RequestURI(requestUrl).Do(context.TODO())
+	if result.Error() != nil {
+		return nil, result.Error()
+	}
+	vm := &kubevirt.VirtualMachine{}
+	err := result.Into(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	return vm, nil
+}
+
+func (p *PoolManager) handleStaleLegacyConfigMapEntries(parentLogger logr.Logger) error {
+	logger := parentLogger.WithName("handleStaleLegacyConfigMapEntries")
+	configMap, err := p.kubeClient.CoreV1().ConfigMaps(p.managerNamespace).Get(context.TODO(), names.WAITING_VMS_CONFIGMAP, metav1.GetOptions{})
+	if err != nil {
+		logger.Error(err, "failed to get config map", "configMapName", names.WAITING_VMS_CONFIGMAP)
+		return err
+	}
+
+	configMapUpdateNeeded := false
+	if configMap.Data == nil {
+		return nil
+	}
+
+	for macAddressDashes, allocationTime := range configMap.Data {
+		// legacy Configmap uses time.RFC3339.
+		t, err := time.Parse(time.RFC3339, allocationTime)
+		if err != nil {
+			// TODO: remove the mac from the wait map??
+			logger.Error(err, "failed to parse allocation time")
 			continue
 		}
 
-		for macAddress, allocationTime := range configMap.Data {
-			t, err := time.Parse(time.RFC3339, allocationTime)
+		logger.Info("data:", "configMapName", names.WAITING_VMS_CONFIGMAP, "configMap.Data", configMap.Data, "macPoolMap", p.macPoolMap)
+
+		if now().After(t.Add(time.Duration(p.waitTime) * time.Second)) {
+			configMapUpdateNeeded = true
+			delete(configMap.Data, macAddressDashes)
+			macAddress := strings.Replace(macAddressDashes, "-", ":", 5)
+			// try to find vm that matches mac and allocate it in macPoolMap
+			err := p.forEachVmInterfaceInClusterRunFunction(func(vmFullName string, iface kubevirt.Interface) {
+				if iface.MacAddress != "" && iface.MacAddress == macAddress {
+					p.macPoolMap.createOrUpdateEntry(macAddress, vmFullName, iface.Name)
+				}
+			})
 			if err != nil {
-				// TODO: remove the mac from the wait map??
-				logger.Error(err, "failed to parse allocation time")
-				continue
+				return err
 			}
 
-			logger.Info("data:", "configMapName", names.WAITING_VMS_CONFIGMAP, "configMap.Data", configMap.Data, "macPoolMap", p.macPoolMap)
-
-			if time.Now().After(t.Add(time.Duration(p.waitTime) * time.Second)) {
-				configMapUpdateNeeded = true
-				delete(configMap.Data, macAddress)
-				macAddress = strings.Replace(macAddress, "-", ":", 5)
-				delete(p.macPoolMap, macAddress)
-				logger.Info("released mac address in waiting loop", "macAddress", macAddress)
+			// if vm was not found, release from macMap
+			if macEntry, exist := p.macPoolMap[macAddress]; exist && macEntry.instanceName == tempVmName {
+				p.macPoolMap.removeMacEntry(macAddress)
```

Comment:
again removing at the same time we iterate is not a good idea

Reply:
removed the function